demos/http3/ossl-nghttp3.c

```diff
@@ -0,0 +1,603 @@
+                                    conn_user_data, stream_user_data);
+
+    SSL_free(stream->s);
+    stream->s = NULL;
+    return ret;
+}
+
+static int h3_conn_reset_stream(nghttp3_conn *h3conn, int64_t stream_id,
+                                uint64_t app_error_code,
+                                void *conn_user_data, void *stream_user_data)
+{
+    int ret = 0;
+    H3_CONN *conn = conn_user_data;
+    H3_STREAM *stream = stream_user_data;
+    SSL_STREAM_RESET_ARGS args = {0};
+
+    if (conn->reset_stream_cb != NULL)
+        ret = conn->reset_stream_cb(h3conn, stream_id, app_error_code,
+                                   conn_user_data, stream_user_data);
+
+    if (stream->s != NULL) {
+        args.quic_error_code = app_error_code;
+
+        if (!SSL_stream_reset(stream->s, &args, sizeof(args)))
+            return 1;
+    }
+
+    return ret;
+}
+
+static int h3_conn_deferred_consume(nghttp3_conn *h3conn, int64_t stream_id,
+                                    size_t consumed,
+                                    void *conn_user_data, void *stream_user_data)
+{
+    int ret = 0;
+    H3_CONN *conn = conn_user_data;
+
+    if (conn->deferred_consume_cb != NULL)
+        ret = conn->deferred_consume_cb(h3conn, stream_id, consumed,
+                                        conn_user_data, stream_user_data);
+
+    conn->consumed_app_data += consumed;
+    return ret;
+}
+
+H3_CONN *H3_CONN_new_for_conn(BIO *qconn_bio,
+                              const nghttp3_callbacks *callbacks,
+                              const nghttp3_settings *settings,
+                              void *user_data)
+{
+    int ec;
+    H3_CONN *conn;
+    H3_STREAM *s_ctl_send = NULL, *s_qpenc_send = NULL, *s_qpdec_send = NULL;
+    nghttp3_settings dsettings = {0};
+    nghttp3_callbacks intl_callbacks = {0};
+    static const unsigned char alpn[] = {2, 'h', '3'};
+
+    if (qconn_bio == NULL) {
+        ERR_raise_data(ERR_LIB_USER, ERR_R_PASSED_NULL_PARAMETER,
+                       "QUIC connection BIO must be provided");
+        return NULL;
+    }
+
+    if ((conn = OPENSSL_zalloc(sizeof(H3_CONN))) == NULL)
+        return NULL;
+
+    conn->qconn_bio = qconn_bio;
+    conn->user_data = user_data;
+
+    if (BIO_get_ssl(qconn_bio, &conn->qconn) == 0) {
+        ERR_raise_data(ERR_LIB_USER, ERR_R_PASSED_INVALID_ARGUMENT,
+                       "BIO must be an SSL BIO");
+        goto err;
+    }
+
+    if ((conn->streams = lh_H3_STREAM_new(h3_stream_hash, h3_stream_eq)) == NULL)
+        goto err;
+
+    if (SSL_in_before(conn->qconn))
+        if (SSL_set_alpn_protos(conn->qconn, alpn, sizeof(alpn))) {
+            /* SSL_set_alpn_protos returns 1 on failure */
+            ERR_raise_data(ERR_LIB_USER, ERR_R_INTERNAL_ERROR,
+                           "failed to configure ALPN");
+            goto err;
+        }
+
+    BIO_set_nbio(conn->qconn_bio, 1);
+
+    if (!SSL_set_default_stream_mode(conn->qconn, SSL_DEFAULT_STREAM_MODE_NONE)) {
```

Comment:
Some extra comments around this line and the `BIO_set_nbio` call above to explain what is going on might be useful